src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs

```diff
@@ -168,25 +169,55 @@ public async Task InitializeAsync()
         }
 
         /// <inheritdoc/>
-        public Task<Transaction> GetTrxAsync(uint256 trxid)
+        public Task<Transaction> GetTransactionByIdAsync(uint256 trxid)
         {
-            return this.blockRepository.GetTrxAsync(trxid);
+            lock (this.blocksCacheLock)
+            {
+                foreach (ChainedHeaderBlock chainedHeaderBlock in this.pendingBlocksCache.Values)
+                {
+                    Transaction tx = chainedHeaderBlock.Block.Transactions.FirstOrDefault(x => x.GetHash() == trxid);
+
+                    if (tx != null)
+                    {
+                        this.logger.LogTrace("Transaction '{0}' was found in the pending blocks cache.", trxid);
+                        return Task.FromResult(tx);
+                    }
+                }
+            }
+
+            return this.blockRepository.GetTransactionByIdAsync(trxid);
         }
 
         /// <inheritdoc/>
-        public Task<uint256> GetTrxBlockIdAsync(uint256 trxid)
+        public Task<uint256> GetBlockIdByTransactionIdAsync(uint256 trxid)
         {
-            return this.blockRepository.GetTrxBlockIdAsync(trxid);
+            lock (this.blocksCacheLock)
```

Comment:
I suggest reverse the order,
First look in disk  (as before), then in memory, to avoid holding the lock unnecessary if not needed.

Reply:
I think this is logical, because the common case will be when we are out of IBD, when we typically serve blocks that are near the tip via the BlockStoreBehavior. These are cached in memory already, so there is no performance hit going to disk.

It will ostensibly be peers trying to sync from genesis that need old uncached blocks, and there would be no point checking the pending cache for these.

Reply:
Discussed on slack. 

@dangershony are you ok with RAM lookup first given the arguments provided on slack?

Reply:
Yes ok this are miner arguments, I also realise there may be a minor race condition if we look at the disk first (and get back null) and between the time we acquire the lock someone is pushing to disk we will have a race condition.

So this is fine